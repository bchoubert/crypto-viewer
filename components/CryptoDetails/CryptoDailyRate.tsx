import React, { FC, useEffect, useRef } from 'react';
import { Animated, StyleSheet, Text, View } from 'react-native';

import Colors from '../../assets/Colors';

import CryptoViewerIconsMap from '../../assets/fonts/baseIcons/CryptoViewerIconsMap';

import UtilsService from '../../services/Utils.service';

const HeightTransitionView = (props) => {
  const translateAnim = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    Animated.timing(
      translateAnim,
      {
        toValue: 1,
        duration: 600,
        useNativeDriver: true,
      }
    ).start();
  }, [translateAnim]);

  const animatedStyle = { transform: [ { scale: translateAnim } ] };

  return (
    <Animated.View style={{ ...props.style, ...animatedStyle }}>
      {props.children}
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  rate_container: {
    height: 40,
    justifyContent: 'flex-start',
    alignItems: 'center',
    marginTop: 25
  },
  rate_plus: {
    padding: 8,
    borderRadius: 17,
    fontSize: 17,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: Colors.white,
    borderWidth: 1,
    borderBottomColor: Colors.white,
    borderTopColor: Colors.white,
    borderLeftColor: Colors.green,
    borderRightColor: Colors.green
  },
  rate_minus: {
    padding: 8,
    borderRadius: 17,
    fontSize: 17,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: Colors.white,
    borderWidth: 1,
    borderBottomColor: Colors.white,
    borderTopColor: Colors.white,
    borderLeftColor: Colors.red,
    borderRightColor: Colors.red
  },
  rate_icon_plus: {
    fontSize: 15,
    paddingRight: 5,
    color: Colors.green,
  },
  rate_icon_minus: {
    fontSize: 15,
    paddingRight: 5,
    color: Colors.red,
  },
  rate_number_plus: {
    color: Colors.green,
  },
  rate_number_minus: {
    color: Colors.red
  },

  cryptoViewerIcon: {
    fontSize: 20,
    fontFamily: 'crypto-viewer'
  }
});

interface CryptoDailyRateProps {
  rate: number;
}

const CryptoDailyRate: FC<CryptoDailyRateProps> = ({
  rate,
}) => {

  return (
    <View style={styles.rate_container}>
      {rate && (
        (rate < 0) ? (
          <HeightTransitionView style={styles.rate_minus}>
            <Text style={{ ...styles.cryptoViewerIcon, ...styles.rate_icon_minus }}>
              {CryptoViewerIconsMap.minus.unicode}
            </Text>
            <Text style={styles.rate_number_minus}>
              {`${UtilsService.truncateNumber(Math.abs(rate))}%`}
            </Text>
          </HeightTransitionView>
        ) : (
          <HeightTransitionView style={styles.rate_plus}>
            <Text style={{ ...styles.cryptoViewerIcon, ...styles.rate_icon_plus }}>
              {CryptoViewerIconsMap.plus.unicode}
            </Text>
            <Text style={styles.rate_number_plus}>
              {`${UtilsService.truncateNumber(Math.abs(rate))}%`}
            </Text>
          </HeightTransitionView>
        )
      )}
    </View>
  );
}

export default CryptoDailyRate;
