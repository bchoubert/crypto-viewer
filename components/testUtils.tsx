/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';

import NavigationProvider from '../contexts/NavigationProvider';
import { SettingsContext } from '../contexts/SettingsProvider';
import ThemeProvider from '../contexts/ThemeProvider';
import TranslationProvider from '../contexts/TranslationProvider';

import { defaultSettings } from '../models/SettingsType';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));
