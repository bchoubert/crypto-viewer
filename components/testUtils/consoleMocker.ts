/* eslint-disable no-console */
const consoleMocker = {
  consoleError: null,

  errorsToSuppress: [
    'Failed prop type',
  ],

  mockConsole: () => {
    consoleMocker.consoleError = console.error;
    console.error = (message: string) => {
      let printMessage = true;

      consoleMocker.errorsToSuppress.forEach((errorToSuppress: string) => {
        if (message.includes(errorToSuppress)) {
          printMessage = false;
        }
      });

      if (printMessage) {
        consoleMocker.consoleError?.(message);
      }
    };
  },

  restoreConsole: () => {
    if (consoleMocker.consoleError) {
      console.error = consoleMocker.consoleError;
    }
  },
};

export default consoleMocker;
