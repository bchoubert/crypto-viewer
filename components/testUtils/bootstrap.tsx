/* eslint-disable no-console */
/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';
import * as PassThoughComponents from '../PassThroughComponents';

import NavigationProvider from '../../contexts/NavigationProvider';
import { SettingsContext } from '../../contexts/SettingsProvider';
import ThemeProvider from '../../contexts/ThemeProvider';
import TranslationProvider from '../../contexts/TranslationProvider';

import { defaultSettings } from '../../models/SettingsType';
import consoleMocker from './consoleMocker';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));

beforeAll(() => {
  consoleMocker.mockConsole();
  // eslint-disable-next-line no-extend-native
  Date.prototype.getTimezoneOffset = jest.fn(() => 0);
});

afterAll(() => {
  consoleMocker.restoreConsole();
  jest.resetAllMocks();
});

jest.spyOn(PassThoughComponents, 'Image');
(PassThoughComponents.Image as unknown as jest.Mock).mockImplementation(() => (<div>IMAGE</div>));
