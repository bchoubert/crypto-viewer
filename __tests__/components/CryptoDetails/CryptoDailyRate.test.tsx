import React from 'react';

import CryptoDailyRate from '../../../components/CryptoDetails/CryptoDailyRate';
import { renderNode } from '../../../components/testUtils/bootstrap';

describe('<CryptoDailyRate />', () => {
  it('Prints correctly', () => {
    const { toJSON } = renderNode(
      <CryptoDailyRate rate={2.5123} />,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('Positive', () => {
    const node = renderNode(
      <CryptoDailyRate rate={2.5123} />,
    );

    node.getByTestId('sign-plus');
    node.getByText('2.51%');
  });

  it('Negative', () => {
    const node = renderNode(
      <CryptoDailyRate rate={-2.5123} />,
    );

    node.getByTestId('sign-minus');
    node.getByText('2.51%');
  });
});
