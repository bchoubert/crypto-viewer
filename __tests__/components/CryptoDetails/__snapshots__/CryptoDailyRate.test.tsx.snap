// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CryptoDailyRate /> Prints correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderRadius": 42,
      "display": "flex",
      "height": 30,
      "justifyContent": "flex-start",
      "marginTop": 25,
      "width": 80,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderRadius": 40,
        "display": "flex",
        "flexDirection": "row",
        "height": 30,
        "justifyContent": "center",
        "width": 80,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#00B344",
          "fontFamily": "crypto-viewer",
          "fontSize": 15,
          "paddingRight": 5,
        }
      }
      testID="sign-plus"
    >
      î ‹
    </Text>
    <Text
      style={
        Object {
          "color": "#00B344",
        }
      }
    >
      2.51%
    </Text>
  </View>
</View>
`;
